src/main/scala/com/fulcrumgenomics/umi/DuplexConsensusCaller.scala

```diff
@@ -0,0 +1,250 @@
+        val filteredAbR2s = filteredYs.filter(_.getSecondOfPairFlag).map(toSourceRead(_, this.minInputBaseQuality))
+        val filteredBaR1s = filteredYs.filter(_.getFirstOfPairFlag).map(toSourceRead(_, this.minInputBaseQuality))
+        val filteredBaR2s = filteredXs.filter(_.getSecondOfPairFlag).map(toSourceRead(_, this.minInputBaseQuality))
+
+        // Call the single-stranded consensus reads
+        val abR1Consensus = ssCaller.consensusCall(filteredAbR1s)
+        val abR2Consensus = ssCaller.consensusCall(filteredAbR2s)
+        val baR1Consensus = ssCaller.consensusCall(filteredBaR1s)
+        val baR2Consensus = ssCaller.consensusCall(filteredBaR2s)
+
+        // Call the duplex reads
+        val duplexR1 = duplexConsensus(abR1Consensus, baR2Consensus, filteredAbR1s ++ filteredBaR2s)
+        val duplexR2 = duplexConsensus(abR2Consensus, baR1Consensus, filteredAbR2s ++ filteredBaR1s)
+
+        // Convert to SAMRecords and return
+        (duplexR1, duplexR2) match {
+          case (Some(r1), Some(r2)) =>
+            Seq(createSamRecord(r1, FirstOfPair), createSamRecord(r2, SecondOfPair))
+          case _                    =>
+            rejectRecords(recs)
+            Nil
+        }
+      case gs =>
+        unreachable("SAMRecords supplied with more than two distinct MI values.")
+    }
+  }
+
+  /**
+    * Constructs a duplex consensus read from a pair of single strand consensus reads.
+    * If either of the incoming reads are undefined, the duplex read will be undefined.
+    *
+    * @param ab the single-strand consensus from one strand
+    * @param ba the single-strand consensus from the other strand
+    * @return a duplex consensus if one can be built
+    */
+  private[umi] def duplexConsensus(ab: Option[VanillaConsensusRead],
+                                   ba: Option[VanillaConsensusRead],
+                                   sourceReads: Seq[SourceRead]): Option[DuplexConsensusRead] = {
+    (ab, ba) match {
+      case (Some(a), Some(b)) =>
+        val len = min(a.length, b.length)
+        val id  = a.id
+        val bases  = new Array[Byte](len)
+        val quals  = new Array[Byte](len)
+        val errors = new Array[Short](len)
+
+        forloop(from=0, until=len) { i =>
+          val aBase = a.bases(i)
+          val bBase = b.bases(i)
+          val aQual = a.quals(i)
+          val bQual = b.quals(i)
+
+          val (base, qual) = {
+            if      (aBase == NoCall || bBase == NoCall) (NoCall, PhredScore.MinValue)
+            else if (aBase == bBase) (aBase, (aQual + bQual).toByte)
+            else if (aQual > bQual)  (aBase, (aQual - bQual).toByte)
+            else if (bQual > aQual)  (bBase, (bQual - aQual).toByte)
+            else (NoCall, PhredScore.MinValue)
+          }
+
+          bases(i)  = base
+          quals(i)  = qual
+          errors(i) = sourceReads.count(s => s.bases(i) != NoCall && s.bases(i) != base).toShort
```

Comment:
I don't think the `toShort` will cap it to 32767 but instead allow for an overflow and become negative.  Can you add a guard here?